### ISSUE TITLE
Some confusion surrounding mixture of float and int operands

### ISSUE THREAD

**Issue by Algebra8** [google/starlark-go#370]
The [specs regarding floating-point numbers](https://github.com/google/starlark-go/blob/master/doc/spec.md#floating-point-numbers) mentions the following:

> An arithmetic operation applied to a mixture of `float` and `int `operands works as if the `int` operand is first converted to a `float`... 

And gives the example of `x // y`, by stating:

> `x // y` yields `floor(x / y)`, that is, the largest integer value not greater than `x / y`. Although the resulting number is integral, it is represented as a `float` if either operand is a `float`.

However, the examples at the bottom of the section include the following:

```python
3.0 // 2.0                                      # 1
```

Where `1 != 1.0`.

 I am assuming the example is incorrect because tests of [`math.ceil` in `testdata/math.star`](https://github.com/google/starlark-go/blob/master/starlark/testdata/math.star#L17) seem to corroborate what the spec states. E.g.:

```python
assert.eq(math.ceil(0), 0.0)
assert.eq(math.ceil(1), 1.0)
assert.eq(math.ceil(10), 10.0)
```

Besides the possible (minor) issue with the example described above, there is something else I am curious/confused about. As far as I can tell, there does not exist any functionality to allow arithmetic operations between `starlark.Int` and `starlark.Float`.

Is the reason for this the portion of the specs that I have pointed out above? Or can there be some situations that call for built-in `float`, `int` operations.

To give you an example, [here](https://github.com/tdakkota/sun/pull/12) I am working on implementing Python's `itertools.count()`  method for starlark go. One issue is the following snippet:

```bash
>>> itertools.count(0.5, 5)
count(0.5, 5)
```

Please note that in Python's implementation, `5` — which is integral, is represented as an `int` and not a `float`.

However, using the [`floatOrInt`](https://github.com/google/starlark-go/blob/master/lib/math/math.go#L111) methodology that `lib/math.go` uses (which can be seen in [`math.ceil()`](https://github.com/google/starlark-go/blob/master/lib/math/math.go#L129)), will yield the following:

```bash
>>> itertools.count(0.5, 5)
count(0.5, 5.0)
```
where `5` has become `5.0`.

To get around the issue I described above, I used the following workaround:

```go
// float or int type to allow mixed inputs.
type floatOrInt struct {
	value starlark.Value
}

// Unpacker for floatOrInt.
func (p *floatOrInt) Unpack(v starlark.Value) error {
	switch v := v.(type) {
	case starlark.Int:
		p.value = v
		return nil
	case starlark.Float:
		p.value = v
		return nil
	}
	return fmt.Errorf("got %s, want float or int", v.Type())
}

func (f *floatOrInt) add(n floatOrInt) error {
	switch _f := f.value.(type) {
	case starlark.Int:
		switch _n := n.value.(type) {
		// int + int
		case starlark.Int:
			f.value = _f.Add(_n)
			return nil
		// int + float
		case starlark.Float:
			_n += _f.Float()
			f.value = _n
			return nil
		}
	case starlark.Float:
		switch _n := n.value.(type) {
		// float + int
		case starlark.Int:
			_f += _n.Float()
			f.value = _f
			return nil
		// float + float
		case starlark.Float:
			_f += _n
			f.value = _f
			return nil
		}
	}

	return fmt.Errorf("error with addition: types are not int, float combos")
}

func (f *floatOrInt) String() string {
	return f.value.String()
}
```

In conclusion, would it make sense to include a `floatOrInt`  struct similar to the `starlark.Int` struct that would be able to handle these kinds of operations, something similar to what I've listed above? Or does this break the rules of the specs described at the top?

A different way to phrase this question is: according to the specs, should `repr(itertools.count(0.5, 5))` return `'count(0.5, 5)'` or `'count(0.5, 5.0)'`

Sorry for the long post and I hope I've explained my situation/confusions clearly.

Thanks!

**Comment by adonovan** [comment]
> However, the examples at the bottom of the section include the following 
> 3.0 // 2.0                                      # 1

The comment should be updated to say 1.0.  (Strictly speaking, 1 is equal to 1.0.)

> Where 1 != 1.0.

In fact 1 is equal 1.0, but I know what you mean: their types differ.


> As far as I can tell, there does not exist any functionality to allow arithmetic operations between starlark.Int and starlark.Float.

The functionality exists, and you already quoted the relevant line of the spec: "An arithmetic operation applied to a mixture of float and int operands works as if the int operand is first converted to a float".  For example. 1.0 + 2 = 3.0.

I think your real question is: how do I write a function that accepts either int or float, and preserves the knowledge of which it got? And you answered it yourself: don't use floatOrInt. I don't think an Unpacker makes sense at all in your case. Just check that the y argument to count(x, y) is int or float, and report an error otherwise. Then invoke the `prev+ y` operation for each iteration.

**Comment by Algebra8** [comment]
Thank you very much for your response 🙏 



> The comment should be updated to say 1.0. (Strictly speaking, 1 is equal to 1.0.)

I can make a PR

> In fact 1 is equal 1.0, but I know what you mean: their types differ.

❤️ 

> The functionality exists, and you already quoted the relevant line of the spec: "An arithmetic operation applied to a mixture of float and int operands works as if the int operand is first converted to a float". For example. 1.0 + 2 = 3.0.

That was silly of me; of course that is the case and you're correct, that functionality does exist. 

The point that I was trying to make was that some Starlark functions, such as `math.ceil`, treat `int`s like `float`s, as can be seen [here](https://github.com/google/starlark-go/blob/master/starlark/testdata/math.star#L17). This is a result of wrapping a `starlark.Int` or a `starlark.Float` in a `float64`. Python's `math.ceil` function does not do that, i.e. in Python`math.ceil(0) == 0`, but in the Starlark-Go implementation `math.ceil(0) == 0.0`.

> I don't think an Unpacker makes sense at all in your case.

Thank you for pointing that out, but with the `intOrFloat` struct that I created it is necessary. Unless you mean that I don't need the `intOrFloat` struct, in which case that's true.

>  If you like, check that the step argument to count(initial, step) is int or float, and report an error otherwise. Then invoke the accum + step operation for each iteration, where accum is initially set to initial.

Thank you very much for the idea, but I'm not sure I completely follow. How would I do the `+` operation if I'm not sure what the types of lhs and rhs are?


> You could even dispense with int-or-float check and just let the plus operation fail if the types are wrong.

Would Go let me compile that?


Without taking too much of your time, I think the point that is most confusing to me is what I mentioned regarding Starlark's `math.ceil` function.

That is, why does `math.ceil(0) == 0.0`, where `type(0.0) == float64`?

Thank you 🙏

**Comment by adonovan** [comment]
> I can make a PR 

Thanks!

> The point that I was trying to make was that some Starlark functions, such as math.ceil, treat ints like floats, as can be seen here. This is a result of wrapping a starlark.Int or a starlark.Float in a float64. Python's math.ceil function does not do that, i.e. in Python math.ceil(0) == 0, but in the Starlark-Go implementation math.ceil(0) == 0.0.

I think you've found a bug: math.ceil should return an int. I suspect it returns a float because it follows the pattern of nearly every other function in the math module, but it should not.

> Thank you for pointing that out, but with the intOrFloat struct that I created it is necessary. Unless you mean that I don't need the intOrFloat struct, in which case that's true.

The reason for the intOrFloat unpacker is that (a) is is used several times in the math module and (b) it obtains a float from an argument of either int or float type. But you don't want to convert int to float: you want to accept either type (and no other), and keep it as is. The easiest way to do that is a simple int-or-float check after UnpackArgs. And if you don't need to do this more than once, you don't need to define a new data type or helper function.

> How would I do the + operation if I'm not sure what the types of lhs and rhs are?

Call `starlark.Binary(PLUS, x, y)`, which computes x+y using Starlark semantics. It fails if the operand types are inappropriate.

> That is, why does math.ceil(0) == 0.0, where type(0.0) == float64?

A bug; see above.

**Comment by Algebra8** [comment]
> Thanks!

Of course :) 

> I think you've found a bug: math.ceil should return an int. I suspect it returns a float because it follows the pattern of nearly every other function in the math module, but it should not.

Is it alright if I take a look, try to fix it, and make another PR for that? I'd love an excuse to dive deeper into the project.

> The easiest way to do that is to call UnpackArgs, and then check that what you got was an int or a float. And if you don't need to do this operation more than once, you don't need a new unpacker/data type/helper function.

That makes a ton of sense; I went a bit overboard with what I did. I'll definitely take this into account and make the necessary changes to the code I wrote. Thanks!

> Call starlark.Binary(PLUS, x, y), which computes x+y using Starlark semantics. It fails if the operand types are inappropriate.

I would not have thought to do that, thank you very much for pointing this out!

**Comment by Algebra8** [comment]
I linked PR #371 to this issue.

I can make a separate issue and PR for the bug you mentioned earlier.

Hope the PR is alright, thanks again :)

**Comment by adonovan** [comment]
> I linked PR #371 to this issue.

Merged. Thanks!

> I can make a separate issue and PR for the bug you mentioned earlier.

Would you? It would be much appreciated.
